ext/afform/core/Civi/Afform/Tokens.php

```diff
@@ -103,70 +105,67 @@ public static function hook_civicrm_tokenValues(GenericHookEvent $e) {
       }
     }
     catch (CryptoException $ex) {
-      \Civi::log()->warning('Civi\Afform\LegacyTokens cannot generate tokens due to crypto exception.', ['exception' => $ex]);
+      \Civi::log()->warning(__CLASS__ . ' cannot generate tokens due to a crypto exception.',
+        ['exception' => $ex]);
     }
   }
 
-  ///**
-  // * Expose tokens for use in UI.
-  // *
-  // * @param \Civi\Token\Event\TokenRegisterEvent $e
-  // */
-  //public static function onRegister(\Civi\Token\Event\TokenRegisterEvent $e) {
-  //  $tokenForms = static::getTokenForms();
-  //  foreach ($tokenForms as $tokenName => $afform) {
-  //    $e->register([
-  //      'entity' => 'afform',
-  //      'field' => $tokenName . 'Url',
-  //      'label' => E::ts('View Form: %1 (URL)', [1 => $afform['title'] ?? $afform['name']]),
-  //    ]);
-  //    $e->register([
-  //      'entity' => 'afform',
-  //      'field' => $tokenName . 'Link',
-  //      'label' => E::ts('View Form: %1 (Full Hyperlink)', [1 => $afform['title'] ?? $afform['name']]),
-  //    ]);
-  //  }
-  //}
-
-  ///**
-  // * Substitute any tokens of the form `{afform.myFormUrl}` or `{afform.myFormLink}` with actual values.
-  // *
-  // * @param \Civi\Token\Event\TokenValueEvent $e
-  // */
-  //public static function onEvaluate(\Civi\Token\Event\TokenValueEvent $e) {
-  //  $activeTokens = $e->getTokenProcessor()->getMessageTokens();
-  //  if (empty($activeTokens['afform'])) {
-  //    return;
-  //  }
-  //
-  //  $tokenForms = static::getTokenForms();
-  //  foreach ($tokenForms as $formName => $afform) {
-  //    if (!array_intersect($activeTokens['afform'], ["{$formName}Url", "{$formName}Link"])) {
-  //      continue;
-  //    }
-  //
-  //    if (empty($afform['server_route'])) {
-  //      \Civi::log()
-  //        ->warning('Civi\Afform\Tokens: Cannot generate link for {formName} -- missing server_route', [
-  //          'formName' => $formName,
-  //        ]);
-  //      continue;
-  //    }
-  //
-  //    foreach ($e->getRows() as $row) {
-  //      /** @var \Civi\Token\TokenRow $row */
-  //      try {
-  //        $url = self::createUrl($afform, $row->context['contactId']);
-  //        $row->format('text/plain')->tokens('afform', "{$formName}Url", $url);
-  //        $row->format('text/html')->tokens('afform', "{$formName}Link",
-  //          sprintf('<a href="%s">%s</a>', htmlentities($url), htmlentities($afform['title'] ?? $afform['name'])));
-  //      }
-  //      catch (CryptoException $e) {
-  //        \Civi::log()->warning('Civi\Afform\Tokens cannot generate tokens due to crypto exception.', ['exception' => $e]);
-  //      }
-  //    }
-  //  }
-  //}
+  public static function listTokens(\Civi\Token\Event\TokenRegisterEvent $e) {
```

Comment:
```suggestion
  public static function listTokens(\Civi\Token\Event\TokenRegisterEvent $e) {
      if (!in_array('contactId', $e->getTokenProcessor()->getContextValues('schema'))) {
      return;
    }
```

Reply:
[As you may have seen](https://chat.civicrm.org/civicrm/pl/88w48y1scfrgfcpg58mf149yzw), I am failing to make the tests fail 😅 on my local environment. But I think this may solve the problem.

We want the tokens to be listed in the "contact" context (the Edit Message Template screen) but not in other contexts that are specific to memberships, cases, etc.

Reply:
Yes, we want the tokens to be listing in the contact context. However, while evaluating there is no contact.

Reply:
Correct.

Reply:
I testing your suggestion.

Reply:
In manual testing, afform submission tokens are still properly evaluated with the suggested change above.

Reply:
Yup, working for me too. However, can't confirm the tests on local (given up). Let's wait for  test build to complete.

Reply:
@highfalutin 

We might have another test failure now `not ok 1185 - Failure: CRM_Core_BAO_MessageTemplateTest::testContactTokens`. Test run is not yet complete: https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/5895/BKPROF=dfl,SUITES=phpunit-crm,label=bknix-tmp/console

Reply:
test run is done and looks like we are back to previous failure: https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/5895/BKPROF=dfl,SUITES=phpunit-crm,label=bknix-tmp/testReport/junit/(root)/CRM_Core_BAO_MessageTemplateTest/testContactTokens/

Reply:
That test will need to be updated. We are getting into a gray area because these are core tests, but they are also testing extension behavior. Yes, core extensions, but extensions.

In this case, I think the best solution is: the test should check for the presence of contact tokens but should not fail if there are additional tokens. @coleman what do you think?

Reply:
@colemanw I tagged you incorrectly the first time above. What do you think?

Reply:
That sounds reasonable @highfalutin - although really @eileenmcnaughton is the person to ask about tokens.

Reply:
Makes sense to me too. @eileenmcnaughton  Let us know your thoughts on this and I can fix the test as a part of this PR.